core/src/main/java/org/apache/carbondata/core/indexstore/columncache/ColumnChunkCache.java

```diff
@@ -0,0 +1,168 @@
+import org.apache.carbondata.core.metadata.schema.table.CarbonTable;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.RemovalListener;
+import com.google.common.cache.RemovalNotification;
+import org.apache.log4j.Logger;
+
+/**
+ * Column Chunk level data cache for table which is enabled for this feature.
+ */
+public class ColumnChunkCache {
+
+  private static Logger LOG =
+      LogServiceFactory.getLogService(ColumnChunkCache.class.getCanonicalName());
+
+  // A global cache for all tables which is enabled for this feature.
+  // The cached object is raw column chunks including raw binary data, so we can
+  // leverage this cache to avoid IO read from underlying storage.
+  private static Map<String, Cache<CacheKey, AbstractRawColumnChunk>> cacheForEachTable =
+      new ConcurrentHashMap<>();
+
+  /**
+   * Enable or disable the column chunk cache for specified table, based on the whether the
+   * corresponding table property is set
+   */
+  public static void setCacheForTable(CarbonTable carbonTable) {
+    String tableId = carbonTable.getTableId();
+    boolean shouldEnable = carbonTable.isColumnCacheEnabled();
+    boolean alreadyEnabled = cacheForEachTable.containsKey(tableId);
+    if (alreadyEnabled && !shouldEnable) {
+      cacheForEachTable.get(tableId).invalidateAll();
+      cacheForEachTable.remove(tableId);
+    } else if (!alreadyEnabled && shouldEnable) {
+      Cache<CacheKey, AbstractRawColumnChunk> cache = createCacheForTable(tableId);
+      cacheForEachTable.put(tableId, cache);
+    }
+  }
+
+  private static Cache<CacheKey, AbstractRawColumnChunk> createCacheForTable(String tableId) {
+    return CacheBuilder.newBuilder()
+        // expire after 10 minutes after last accessed
+        .expireAfterAccess(10, TimeUnit.MINUTES)
+        // max number of entries
+        .maximumSize(1000)
+        .removalListener(new RemovalListener<CacheKey, AbstractRawColumnChunk>() {
+          @Override
+          public void onRemoval(RemovalNotification<CacheKey, AbstractRawColumnChunk> entry) {
+            AbstractRawColumnChunk chunk = entry.getValue();
+            if (chunk != null) {
+              chunk.setCached(false);
+              chunk.freeMemory();
+              if (LOG.isInfoEnabled()) {
+                CacheKey key = entry.getKey();
+                LOG.info(String.format("column cache removed: key %s", key.toString()));
+              }
+            }
+          }
+        }).build();
+  }
+
+  /**
+   * Return true if column chunk cache is enabled for specified table
+   */
+  public static boolean isEnabledForTable(String tableId) {
+    return tableId != null && cacheForEachTable.containsKey(tableId);
+  }
+
+  /**
+   * Put column chunk into the cache
+   */
+  public static void put(String tableId, CacheKey key, AbstractRawColumnChunk chunk) {
+    if (isEnabledForTable(tableId)) {
+      cacheForEachTable.get(tableId).put(key, chunk);
```

Comment:
Concurrent query performance will be impacted with this as this is a system-level cache. concurrent hash map will have a lock.